// node-opcua-address-space/dist/src/nodeset_tools/nodeset_to_xml.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dumpXml = void 0;
// tslint:disable:no-var-requires
const XMLWriter = require("xml-writer");
const node_opcua_assert_1 = require("node-opcua-assert");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_data_model_2 = require("node-opcua-data-model");
const node_opcua_factory_1 = require("node-opcua-factory");
const node_opcua_factory_2 = require("node-opcua-factory");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const utils = require("node-opcua-utils");
const node_opcua_variant_1 = require("node-opcua-variant");
const node_opcua_variant_2 = require("node-opcua-variant");
const node_opcua_variant_3 = require("node-opcua-variant");
const node_opcua_basic_types_1 = require("node-opcua-basic-types");
const node_opcua_types_1 = require("node-opcua-types");
const base_node_1 = require("../base_node");
const namespace_1 = require("../namespace");
const reference_1 = require("../reference");
const ua_data_type_1 = require("../ua_data_type");
const ua_method_1 = require("../ua_method");
const ua_object_1 = require("../ua_object");
const ua_object_type_1 = require("../ua_object_type");
const ua_reference_type_1 = require("../ua_reference_type");
const ua_variable_1 = require("../ua_variable");
const ua_variable_type_1 = require("../ua_variable_type");
const construct_namespace_dependency_1 = require("./construct_namespace_dependency");
function _hash(node) {
    return node.nodeId.toString();
}
function _dumpDisplayName(xw, node) {
    if (node.displayName && node.displayName[0]) {
        xw.startElement("DisplayName").text(node.displayName[0].text).endElement();
    }
}
function _dumpDescription(xw, node) {
    if (node.description) {
        let desc = node.description.text;
        desc = desc || "";
        xw.startElement("Description").text(desc).endElement();
    }
}
function translateNodeId(xw, nodeId) {
    node_opcua_assert_1.assert(nodeId instanceof node_opcua_nodeid_1.NodeId);
    const nn = xw.translationTable[nodeId.namespace];
    const translatedNode = new node_opcua_nodeid_1.NodeId(nodeId.identifierType, nodeId.value, nn);
    return translatedNode;
}
function n(xw, nodeId) {
    return translateNodeId(xw, nodeId).toString().replace("ns=0;", "");
}
function translateBrowseName(xw, browseName) {
    node_opcua_assert_1.assert(browseName instanceof node_opcua_data_model_2.QualifiedName);
    const nn = xw.translationTable[browseName.namespaceIndex];
    const translatedBrowseName = new node_opcua_data_model_2.QualifiedName({ namespaceIndex: nn, name: browseName.name });
    return translatedBrowseName;
}
function b(xw, browseName) {
    return translateBrowseName(xw, browseName).toString().replace("ns=0;", "");
}
function _dumpReferences(xw, node) {
    xw.startElement("References");
    const addressSpace = node.addressSpace;
    const aggregateReferenceType = addressSpace.findReferenceType("Aggregates");
    const hasChildReferenceType = addressSpace.findReferenceType("HasChild");
    const hasSubtypeReferenceType = addressSpace.findReferenceType("HasSubtype");
    const hasTypeDefinitionReferenceType = addressSpace.findReferenceType("HasTypeDefinition");
    const nonHierarchicalReferencesType = addressSpace.findReferenceType("NonHierarchicalReferences");
    const organizesReferencesType = addressSpace.findReferenceType("Organizes");
    const connectsToReferenceType = addressSpace.findReferenceType("ConnectsTo");
    const hasEventSourceReferenceType = addressSpace.findReferenceType("HasEventSource");
    function referenceToKeep(reference) {
        const referenceType = reference._referenceType;
        // get the direct backward reference to a external namespace
        if (referenceType.isSupertypeOf(aggregateReferenceType) && !reference.isForward) {
            if (reference.nodeId.namespace !== node.nodeId.namespace) {
                // todo: may be check that reference.nodeId.namespace is one of the namespace
                // on which our namespace is build and not a derived one !
                // xx console.log("xxxxxxxxxxxxxx Keeping => ", referenceType.toString(), reference.node?.nodeId.toString());
                return true;
            }
        }
        // only keep
        if (referenceType.isSupertypeOf(aggregateReferenceType) && reference.isForward) {
            return true;
        }
        else if (referenceType.isSupertypeOf(hasSubtypeReferenceType) && !reference.isForward) {
            return true;
        }
        else if (referenceType.isSupertypeOf(hasTypeDefinitionReferenceType) && reference.isForward) {
            return true;
        }
        else if (referenceType.isSupertypeOf(nonHierarchicalReferencesType) && reference.isForward) {
            return true;
        }
        else if (referenceType.isSupertypeOf(organizesReferencesType) && !reference.isForward) {
            return true;
        }
        else if (connectsToReferenceType && referenceType.isSupertypeOf(connectsToReferenceType) && reference.isForward) {
            return true;
        }
        else if (referenceType.isSupertypeOf(hasEventSourceReferenceType) && reference.isForward) {
            return true;
        }
        return false;
    }
    const references = node.allReferences().filter(referenceToKeep);
    for (const reference of references) {
        if (reference._referenceType.browseName.toString() === "HasSubtype" && reference.isForward) {
            continue;
        }
        // only output inverse Reference
        xw.startElement("Reference");
        xw.writeAttribute("ReferenceType", b(xw, reference._referenceType.browseName));
        if (!reference.isForward) {
            xw.writeAttribute("IsForward", reference.isForward ? "true" : "false");
        }
        xw.text(n(xw, reference.nodeId));
        xw.endElement();
    }
    xw.endElement();
}
function _dumpLocalizedText(xw, v) {
    xw.startElement("Locale");
    if (v.locale) {
        xw.text(v.locale);
    }
    xw.endElement();
    xw.startElement("Text");
    if (v.text) {
        xw.text(v.text);
    }
    xw.endElement();
}
/*
<uax:ExtensionObject>
    <uax:TypeId>
        <uax:Identifier>i=339</uax:Identifier>
    </uax:TypeId>
    <uax:Body>
        <BuildInfo xmlns="http://opcfoundation.org/UA/2008/02/Types.xsd">
            <ProductUri></ProductUri>
            <ManufacturerName></ManufacturerName>
            <ProductName></ProductName>
            <SoftwareVersion></SoftwareVersion>
            <BuildNumber></BuildNumber>
            <BuildDate>1900-01-01T00:00:00Z</BuildDate>
        </BuildInfo>
    </uax:Body>
</uax:ExtensionObject>
*/
function _dumpExtensionObject(xw, v) {
    if (!v) {
        return;
    }
    xw.startElement("TypeId");
    _dumpNodeId(xw, v.schema.encodingDefaultXml);
    xw.endElement();
    xw.startElement("Body");
    xw.endElement();
}
function _dumpNodeId(xw, v) {
    xw.startElement("Identifier");
    xw.text(v.toString());
    xw.endElement();
}
// tslint:disable:no-console
function _dumpVariantValue(xw, dataType, value) {
    if (value === undefined || value === null) {
        return;
    }
    switch (dataType) {
        case node_opcua_variant_3.DataType.Null:
            break;
        case node_opcua_variant_3.DataType.LocalizedText:
            xw.startElement(node_opcua_variant_3.DataType[dataType]);
            // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            _dumpLocalizedText(xw, value);
            xw.endElement();
            break;
        case node_opcua_variant_3.DataType.NodeId:
            xw.startElement(node_opcua_variant_3.DataType[dataType]);
            // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            _dumpNodeId(xw, value);
            xw.endElement();
            break;
        case node_opcua_variant_3.DataType.DateTime:
            xw.startElement(node_opcua_variant_3.DataType[dataType]);
            // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            xw.text(value.toISOString());
            xw.endElement();
            break;
        case node_opcua_variant_3.DataType.Int64:
        case node_opcua_variant_3.DataType.UInt64:
            xw.startElement(node_opcua_variant_3.DataType[dataType]);
            // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            xw.text(value[1].toString());
            xw.endElement();
            break;
        case node_opcua_variant_3.DataType.Boolean:
        case node_opcua_variant_3.DataType.SByte:
        case node_opcua_variant_3.DataType.Byte:
        case node_opcua_variant_3.DataType.Float:
        case node_opcua_variant_3.DataType.Double:
        case node_opcua_variant_3.DataType.Int16:
        case node_opcua_variant_3.DataType.Int32:
        case node_opcua_variant_3.DataType.UInt16:
        case node_opcua_variant_3.DataType.UInt32:
        case node_opcua_variant_3.DataType.String:
            if (value !== undefined && value !== null) {
                xw.startElement(node_opcua_variant_3.DataType[dataType]);
                // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
                xw.text(value.toString());
                xw.endElement();
            }
            break;
        case node_opcua_variant_3.DataType.ByteString:
            if (value !== undefined && value !== null) {
                xw.startElement(node_opcua_variant_3.DataType[dataType]);
                // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
                const base64 = value.toString("base64");
                xw.text(base64.match(/.{0,80}/g).join("\n"));
                xw.endElement();
            }
            break;
        case node_opcua_variant_3.DataType.Guid:
            /*
             <uax:Guid>
                <uax:String>947c29a7-490d-4dc9-adda-1109e3e8fcb7</uax:String>
            </uax:Guid>
            */
            if (value !== undefined && value !== null) {
                xw.startElement(node_opcua_variant_3.DataType[dataType]);
                // xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
                xw.startElement("String");
                xw.text(value.toString());
                xw.endElement();
                xw.endElement();
            }
            break;
        case node_opcua_variant_3.DataType.ExtensionObject:
            xw.startElement(node_opcua_variant_3.DataType[dataType]);
            _dumpExtensionObject(xw, value);
            xw.endElement();
            break;
        case node_opcua_variant_3.DataType.QualifiedName:
        case node_opcua_variant_3.DataType.StatusCode:
        default:
            throw new Error("_dumpVariantValue!! incomplete  dataType=" + dataType + " - v=" + node_opcua_variant_3.DataType[dataType] + "  value = " + value);
    }
}
// tslint:disable:no-console
function _dumpVariantInnerValue(xw, dataType, value) {
    switch (dataType) {
        case null:
        case node_opcua_variant_3.DataType.Null:
            break;
        case node_opcua_variant_3.DataType.LocalizedText:
            _dumpLocalizedText(xw, value);
            break;
        case node_opcua_variant_3.DataType.NodeId:
            _dumpNodeId(xw, value);
            break;
        case node_opcua_variant_3.DataType.DateTime:
            xw.text(value.toISOString());
            break;
        case node_opcua_variant_3.DataType.Int64:
        case node_opcua_variant_3.DataType.UInt64:
            xw.text(value[1].toString());
            break;
        case node_opcua_variant_3.DataType.Boolean:
        case node_opcua_variant_3.DataType.Byte:
        case node_opcua_variant_3.DataType.Float:
        case node_opcua_variant_3.DataType.Double:
        case node_opcua_variant_3.DataType.Int16:
        case node_opcua_variant_3.DataType.Int32:
        case node_opcua_variant_3.DataType.UInt16:
        case node_opcua_variant_3.DataType.UInt32:
        case node_opcua_variant_3.DataType.String:
            xw.text(value.toString());
            break;
        case node_opcua_variant_3.DataType.ByteString:
        case node_opcua_variant_3.DataType.QualifiedName:
        case node_opcua_variant_3.DataType.StatusCode:
        default:
            throw new Error("_dumpVariantInnerValue incomplete " + value + " " + "DataType=" + dataType + "=" + node_opcua_variant_3.DataType[dataType]);
    }
}
/**
 *
 * @param field
 */
function findBaseDataType(field) {
    if (field.fieldType === "UAString") {
        return node_opcua_variant_3.DataType.String;
    }
    const result = node_opcua_variant_3.DataType[field.fieldType];
    if (!result) {
        throw new Error("cannot find baseDataType of " + field.name + "= " + field.fieldType);
    }
    return result;
}
/**
 *
 * @param xw
 * @param schema
 * @param value
 * @private
 */
function _dumpVariantExtensionObjectValue_Body(xw, schema, value) {
    if (value) {
        xw.startElement(schema.name);
        if (value) {
            for (const field of schema.fields) {
                const v = value[field.name];
                if (v !== null && v !== undefined) {
                    xw.startElement(utils.capitalizeFirstLetter(field.name));
                    try {
                        const baseType = findBaseDataType(field);
                        _dumpVariantInnerValue(xw, baseType, v);
                    }
                    catch (err) {
                        console.log("Error in _dumpVariantExtensionObjectValue_Body !!!", err.message);
                        console.log(schema.name);
                        console.log(field);
                        // throw err;
                    }
                    xw.endElement();
                }
            }
        }
        xw.endElement();
    }
}
/* encode object as XML */
function _dumpVariantExtensionObjectValue(xw, schema, value) {
    xw.startElement("ExtensionObject");
    {
        xw.startElement("TypeId");
        {
            // find HasEncoding node
            const encodingDefaultXml = node_opcua_factory_1.getStructureTypeConstructor(schema.name).encodingDefaultXml;
            if (!encodingDefaultXml) {
                console.log("?????");
            }
            // xx var encodingDefaultXml = schema.encodingDefaultXml;
            xw.startElement("Identifier");
            xw.text(encodingDefaultXml.toString());
            xw.endElement();
        }
        xw.endElement();
        xw.startElement("Body");
        _dumpVariantExtensionObjectValue_Body(xw, schema, value);
        xw.endElement();
    }
    xw.endElement();
}
function _isDefaultValue(value) {
    // detect default value
    if (value.arrayType === node_opcua_variant_2.VariantArrayType.Scalar) {
        switch (value.dataType) {
            case node_opcua_variant_3.DataType.ExtensionObject:
                if (!value.value) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.DateTime:
                if (!value.value || value.value.getTime() === node_opcua_basic_types_1.minOPCUADate) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.ByteString:
                if (!value.value || value.value.length === 0) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.Boolean:
                if (!value.value) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.SByte:
            case node_opcua_variant_3.DataType.Byte:
            case node_opcua_variant_3.DataType.UInt16:
            case node_opcua_variant_3.DataType.UInt32:
            case node_opcua_variant_3.DataType.Int16:
            case node_opcua_variant_3.DataType.Int32:
            case node_opcua_variant_3.DataType.Double:
            case node_opcua_variant_3.DataType.Float:
                if (value.value === 0 || value.value === null) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.String:
                if (value.value === null || value.value === "") {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.UInt64:
            case node_opcua_variant_3.DataType.UInt64:
                if (0 === coerceInt64ToInt32(value.value)) {
                    return true;
                }
                break;
            case node_opcua_variant_3.DataType.LocalizedText:
                if (!value.value) {
                    return true;
                }
                const l = value.value;
                if (!l.locale && !l.text) {
                    return true;
                }
                break;
        }
        return false;
    }
    else {
        if (!value.value || value.value.length === 0) {
            return true;
        }
        return false;
    }
}
function _dumpValue(xw, node, value) {
    const addressSpace = node.addressSpace;
    // istanbul ignore next
    if (value === null || value === undefined) {
        return;
    }
    node_opcua_assert_1.assert(value instanceof node_opcua_variant_1.Variant);
    const dataTypeNode = addressSpace.findNode(node.dataType);
    if (!dataTypeNode) {
        console.log("Cannot find dataType:", node.dataType);
        return;
    }
    const dataTypeName = dataTypeNode.browseName.name.toString();
    const baseDataTypeName = node_opcua_variant_3.DataType[value.dataType];
    if (baseDataTypeName === "Null") {
        return;
    }
    node_opcua_assert_1.assert(typeof baseDataTypeName === "string");
    // determine if dataTypeName is a ExtensionObject
    const isExtensionObject = value.dataType === node_opcua_variant_3.DataType.ExtensionObject;
    if (_isDefaultValue(value)) {
        return;
    }
    xw.startElement("Value");
    if (isExtensionObject) {
        if (node_opcua_factory_1.hasStructuredType(dataTypeName)) {
            const schema = node_opcua_factory_2.getStructuredTypeSchema(dataTypeName);
            const encodeXml = _dumpVariantExtensionObjectValue.bind(null, xw, schema);
            if (value.arrayType === node_opcua_variant_2.VariantArrayType.Array) {
                xw.startElement("ListOf" + baseDataTypeName);
                value.value.forEach(encodeXml);
                xw.endElement();
            }
            else if (value.arrayType === node_opcua_variant_2.VariantArrayType.Scalar) {
                encodeXml(value.value);
            }
            else {
                throw new Error("Unsupported case");
            }
        }
    }
    else {
        const encodeXml = _dumpVariantValue.bind(null, xw, value.dataType);
        if (value.arrayType === node_opcua_variant_2.VariantArrayType.Matrix) {
            console.log("Warning _dumpValue : Matrix not supported yet");
            xw.startElement("ListOf" + dataTypeName);
            xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            value.value.forEach(encodeXml);
            xw.endElement();
        }
        else if (value.arrayType === node_opcua_variant_2.VariantArrayType.Array) {
            xw.startElement("ListOf" + dataTypeName);
            xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2008/02/Types.xsd");
            value.value.forEach(encodeXml);
            xw.endElement();
        }
        else if (value.arrayType === node_opcua_variant_2.VariantArrayType.Scalar) {
            encodeXml(value.value);
        }
        else {
            throw new Error("Unsupported case");
        }
    }
    xw.endElement();
}
function _dumpArrayDimensionsAttribute(xw, node) {
    if (node.arrayDimensions) {
        if (node.arrayDimensions.length === 1 && node.arrayDimensions[0] === 0) {
            return;
        }
        xw.writeAttribute("ArrayDimensions", node.arrayDimensions.join(","));
    }
}
function visitUANode(node, options, forward) {
    node_opcua_assert_1.assert(typeof forward === "boolean");
    const addressSpace = node.addressSpace;
    options.elements = options.elements || [];
    options.index_el = options.index_el || {};
    // visit references
    function process_reference(reference) {
        //  only backward or forward references
        if (reference.isForward !== forward) {
            return;
        }
        if (reference.nodeId.namespace === 0) {
            return; // skip OPCUA namespace
        }
        const k = _hash(reference);
        if (!options.index_el[k]) {
            options.index_el[k] = 1;
            const o = addressSpace.findNode(k);
            if (o) {
                visitUANode(o, options, forward);
            }
        }
    }
    node.ownReferences().forEach(process_reference);
    options.elements.push(node);
    return node;
}
function dumpReferencedNodes(xw, node, forward) {
    const addressSpace = node.addressSpace;
    if (!forward) {
        {
            const r = node.findReferencesEx("HasTypeDefinition");
            if (r && r.length) {
                node_opcua_assert_1.assert(r.length === 1);
                const typeDefinitionObj = reference_1.Reference.resolveReferenceNode(addressSpace, r[0]);
                if (!typeDefinitionObj) {
                    console.log(node.toString());
                    console.log("Warning : " + node.browseName.toString() + " unknown typeDefinition, ", r[0].toString());
                }
                else {
                    node_opcua_assert_1.assert(typeDefinitionObj instanceof base_node_1.BaseNode);
                    if (typeDefinitionObj.nodeId.namespace === node.nodeId.namespace) {
                        // only output node if it is on the same namespace
                        if (!xw.visitedNode[_hash(typeDefinitionObj)]) {
                            typeDefinitionObj.dumpXML(xw);
                        }
                    }
                }
            }
        }
        //
        {
            const r = node.findReferencesEx("HasSubtype", node_opcua_data_model_1.BrowseDirection.Inverse);
            if (r && r.length) {
                const subTypeOf = reference_1.Reference.resolveReferenceNode(addressSpace, r[0]);
                node_opcua_assert_1.assert(r.length === 1);
                if (subTypeOf.nodeId.namespace === node.nodeId.namespace) {
                    // only output node if it is on the same namespace
                    if (!xw.visitedNode[_hash(subTypeOf)]) {
                        subTypeOf.dumpXML(xw);
                    }
                }
            }
        }
    }
    else {
        const r = node.findReferencesEx("Aggregates", node_opcua_data_model_1.BrowseDirection.Forward);
        for (const reference of r) {
            const nodeChild = reference_1.Reference.resolveReferenceNode(addressSpace, reference);
            node_opcua_assert_1.assert(nodeChild instanceof base_node_1.BaseNode);
            if (nodeChild.nodeId.namespace === node.nodeId.namespace) {
                if (!xw.visitedNode[_hash(nodeChild)]) {
                    console.log(node.nodeId.toString(), " dumping child ", nodeChild.browseName.toString(), nodeChild.nodeId.toString());
                    nodeChild.dumpXML(xw);
                }
            }
        }
    }
}
const currentReadFlag = node_opcua_data_model_1.makeAccessLevelFlag("CurrentRead");
function dumpCommonAttributes(xw, node) {
    xw.writeAttribute("NodeId", n(xw, node.nodeId));
    xw.writeAttribute("BrowseName", b(xw, node.browseName));
    if (node.hasOwnProperty("symbolicName")) {
        xw.writeAttribute("SymbolicName", node.symbolicName);
    }
    if (node.hasOwnProperty("isAbstract")) {
        if (node.isAbstract) {
            xw.writeAttribute("IsAbstract", node.isAbstract ? "true" : "false");
        }
    }
    if (node.hasOwnProperty("accessLevel")) {
        // CurrentRead is by default
        if (node.accessLevel !== currentReadFlag) {
            xw.writeAttribute("AccessLevel", node.accessLevel);
        }
    }
}
function dumpCommonElements(xw, node) {
    _dumpDisplayName(xw, node);
    _dumpDescription(xw, node);
    _dumpReferences(xw, node);
}
function coerceInt64ToInt32(int64) {
    if (typeof int64 === "number") {
        return int64;
    }
    console.log(int64);
    if (int64[0] === 4294967295 && int64[1] === 4294967295) {
        return 0xffffffff;
    }
    node_opcua_assert_1.assert(int64[0] === 0, "???");
    return int64[1];
}
function _dumpEnumDefinition(xw, enumDefinition) {
    enumDefinition.fields = enumDefinition.fields || [];
    for (const defItem of enumDefinition.fields) {
        xw.startElement("Field");
        xw.writeAttribute("Name", defItem.name);
        if (!utils.isNullOrUndefined(defItem.value)) {
            xw.writeAttribute("Value", coerceInt64ToInt32(defItem.value));
        }
        if (defItem.description && defItem.description.text) {
            xw.startElement("Description");
            xw.text(defItem.description.text.toString());
            xw.endElement();
        }
        xw.endElement();
    }
}
function _dumpStructureDefinition(xw, structureDefinition) {
    /*
     * note: baseDataType and defaultEncodingId are implicit and not stored in the XML file ??
     *
     */
    const baseDataType = structureDefinition.baseDataType;
    const defaultEncodingId = structureDefinition.defaultEncodingId;
    structureDefinition.fields = structureDefinition.fields || [];
    for (const defItem /*: StructureField*/ of structureDefinition.fields) {
        xw.startElement("Field");
        xw.writeAttribute("Name", defItem.name);
        if (defItem.arrayDimensions) {
            xw.writeAttribute("ArrayDimensions", defItem.arrayDimensions.map((x) => x.toString()).join(","));
        }
        if (defItem.valueRank !== undefined && defItem.valueRank !== -1) {
            xw.writeAttribute("ValueRank", defItem.valueRank);
        }
        if (defItem.isOptional /* && defItem.isOptional !== false */) {
            xw.writeAttribute("IsOptional", defItem.isOptional.toString());
        }
        if (defItem.maxStringLength !== undefined && defItem.maxStringLength !== 0) {
            xw.writeAttribute("MaxStringLength", defItem.maxStringLength);
        }
        // todo : SymbolicName ( see AutoId )
        if (defItem.dataType) {
            // todo : namespace translation !
            xw.writeAttribute("DataType", n(xw, defItem.dataType));
        }
        if (defItem.description && defItem.description.text) {
            xw.startElement("Description");
            xw.text(defItem.description.text.toString());
            xw.endElement();
        }
        xw.endElement();
    }
}
function _dumpUADataTypeDefinition(xw, node) {
    const $definition = node._getDefinition();
    if (!$definition) {
        return;
    }
    if ($definition instanceof node_opcua_types_1.EnumDefinition) {
        xw.startElement("Definition");
        xw.writeAttribute("Name", node.browseName.name);
        _dumpEnumDefinition(xw, $definition);
        xw.endElement();
        return;
    }
    if ($definition instanceof node_opcua_types_1.StructureDefinition) {
        xw.startElement("Definition");
        xw.writeAttribute("Name", node.browseName.name);
        if ($definition.structureType === node_opcua_types_1.StructureType.Union) {
            xw.writeAttribute("IsUnion", "true");
        }
        /*
        // todo
        $definition.baseDataType;
        // todo
        $definition.structureType;
        */
        _dumpStructureDefinition(xw, $definition);
        xw.endElement();
        return;
    }
    // throw new Error("_dumpUADataTypeDefinition: Should not get here !");
}
function dumpUADataType(xw, node) {
    _markAsVisited(xw, node);
    xw.startElement("UADataType");
    xw.writeAttribute("NodeId", n(xw, node.nodeId));
    xw.writeAttribute("BrowseName", b(xw, node.browseName));
    if (node.symbolicName !== node.browseName.name) {
        xw.writeAttribute("SymbolicName", node.symbolicName);
    }
    if (node.isAbstract) {
        xw.writeAttribute("IsAbstract", node.isAbstract ? "true" : "false");
    }
    _dumpDisplayName(xw, node);
    _dumpReferences(xw, node);
    _dumpUADataTypeDefinition(xw, node);
    xw.endElement();
    dumpAggregates(xw, node);
}
ua_data_type_1.UADataType.prototype.dumpXML = function (xw) {
    dumpUADataType(xw, this);
};
function _markAsVisited(xw, node) {
    xw.visitedNode = xw.visitedNode || {};
    node_opcua_assert_1.assert(!xw.visitedNode[_hash(node)]);
    xw.visitedNode[_hash(node)] = 1;
}
function dumpUAVariable(xw, node) {
    _markAsVisited(xw, node);
    dumpReferencedNodes(xw, node, false);
    const addressSpace = node.addressSpace;
    xw.startElement("UAVariable");
    {
        // attributes
        dumpCommonAttributes(xw, node);
        if (node.valueRank !== -1) {
            // -1 = Scalar
            xw.writeAttribute("ValueRank", node.valueRank);
        }
        _dumpArrayDimensionsAttribute(xw, node);
        const dataTypeNode = addressSpace.findNode(node.dataType);
        if (dataTypeNode) {
            // verify that data Type is in alias
            // xx const dataTypeName = dataTypeNode.browseName.toString();
            const dataTypeName = b(xw, resolveDataTypeName(addressSpace, dataTypeNode.nodeId));
            xw.writeAttribute("DataType", dataTypeName);
        }
    }
    {
        // sub elements
        dumpCommonElements(xw, node);
        _dumpValue(xw, node, node.readValue().value);
    }
    xw.endElement();
    dumpAggregates(xw, node);
}
ua_variable_1.UAVariable.prototype.dumpXML = function (xw) {
    dumpUAVariable(xw, this);
};
ua_reference_type_1.UAReferenceType.prototype.dumpXML = function (xw) {
    dumpReferenceType(xw, this);
};
function dumpUAVariableType(xw, node) {
    xw.visitedNode = xw.visitedNode || {};
    node_opcua_assert_1.assert(!xw.visitedNode[_hash(node)]);
    xw.visitedNode[_hash(node)] = 1;
    dumpReferencedNodes(xw, node, false);
    const addressSpace = node.addressSpace;
    xw.startElement("UAVariableType");
    {
        // attributes
        dumpCommonAttributes(xw, node);
        if (node.valueRank !== -1) {
            xw.writeAttribute("ValueRank", node.valueRank);
        }
        const dataTypeNode = addressSpace.findNode(node.dataType);
        if (!dataTypeNode) {
            // throw new Error(" cannot find datatype " + node.dataType);
            console.log(" cannot find datatype " +
                node.dataType +
                " for node " +
                node.browseName.toString() +
                " id =" +
                node.nodeId.toString());
        }
        else {
            const dataTypeName = b(xw, resolveDataTypeName(addressSpace, dataTypeNode.nodeId));
            xw.writeAttribute("DataType", dataTypeName);
        }
    }
    {
        _dumpArrayDimensionsAttribute(xw, node);
        // sub elements
        dumpCommonElements(xw, node);
        _dumpValue(xw, node, node.value);
    }
    xw.endElement();
    dumpAggregates(xw, node);
}
ua_variable_type_1.UAVariableType.prototype.dumpXML = function (xw) {
    dumpUAVariableType(xw, this);
};
function dumpUAObject(xw, node) {
    xw.writeComment("Object - " + b(xw, node.browseName) + " {{{{ ");
    xw.visitedNode = xw.visitedNode || {};
    node_opcua_assert_1.assert(!xw.visitedNode[_hash(node)]);
    xw.visitedNode[_hash(node)] = 1;
    // dump SubTypeOf and HasTypeDefinition
    dumpReferencedNodes(xw, node, false);
    xw.startElement("UAObject");
    dumpCommonAttributes(xw, node);
    dumpCommonElements(xw, node);
    xw.endElement();
    // dump aggregates nodes ( Properties / components )
    dumpAggregates(xw, node);
    dumpElementInFolder(xw, node);
    xw.writeComment("Object - " + b(xw, node.browseName) + " }}}} ");
}
ua_object_1.UAObject.prototype.dumpXML = function (xw) {
    dumpUAObject(xw, this);
};
function dumpElementInFolder(xw, node) {
    const aggregates = node
        .getFolderElements()
        .sort((x, y) => (x.browseName.name.toString() > y.browseName.name.toString() ? 1 : -1));
    for (const aggregate of aggregates) {
        // do not export node that do not belong to our namespace
        if (node.nodeId.namespace !== aggregate.nodeId.namespace) {
            return;
        }
        if (!xw.visitedNode[_hash(aggregate)]) {
            aggregate.dumpXML(xw);
        }
    }
}
function dumpAggregates(xw, node) {
    // Xx xw.writeComment("Aggregates {{ ");
    const aggregates = node
        .getAggregates()
        .sort((x, y) => (x.browseName.name.toString() > y.browseName.name.toString() ? 1 : -1));
    for (const aggregate of aggregates) {
        // do not export node that do not belong to our namespace
        if (node.nodeId.namespace !== aggregate.nodeId.namespace) {
            return;
        }
        if (!xw.visitedNode[_hash(aggregate)]) {
            aggregate.dumpXML(xw);
        }
    }
    // Xx xw.writeComment("Aggregates }} ");
}
function dumpUAObjectType(xw, node) {
    node_opcua_assert_1.assert(node instanceof ua_object_type_1.UAObjectType);
    xw.writeComment("ObjectType - " + b(xw, node.browseName) + " {{{{ ");
    _markAsVisited(xw, node);
    // dump SubtypeOf and HasTypeDefinition
    dumpReferencedNodes(xw, node, false);
    xw.startElement("UAObjectType");
    dumpCommonAttributes(xw, node);
    dumpCommonElements(xw, node);
    xw.endElement();
    dumpAggregates(xw, node);
    xw.writeComment("ObjectType - " + b(xw, node.browseName) + " }}}}");
}
ua_object_type_1.UAObjectType.prototype.dumpXML = function (xw) {
    dumpUAObjectType(xw, this);
};
function dumpUAMethod(xw, node) {
    _markAsVisited(xw, node);
    dumpReferencedNodes(xw, node, false);
    xw.startElement("UAMethod");
    dumpCommonAttributes(xw, node);
    if (node.methodDeclarationId) {
        xw.writeAttribute("MethodDeclarationId", n(xw, node.methodDeclarationId));
    }
    dumpCommonElements(xw, node);
    xw.endElement();
    dumpAggregates(xw, node);
}
ua_method_1.UAMethod.prototype.dumpXML = function (xw) {
    dumpUAMethod(xw, this);
};
function resolveDataTypeName(addressSpace, dataType) {
    let dataTypeNode = null;
    // istanbul ignore next
    if (typeof dataType === "string") {
        dataTypeNode = addressSpace.findDataType(dataType);
    }
    else {
        node_opcua_assert_1.assert(dataType instanceof node_opcua_nodeid_1.NodeId);
        const o = addressSpace.findNode(dataType.toString());
        dataTypeNode = o ? o : null;
    }
    if (!dataTypeNode) {
        throw new Error("Cannot find dataTypeName " + dataType);
    }
    return dataTypeNode.browseName;
}
function buildUpAliases(node, xw, options) {
    const addressSpace = node.addressSpace;
    options.aliases = options.aliases || {};
    options.aliases_visited = options.aliases_visited || {};
    const k = _hash(node);
    // istanbul ignore next
    if (options.aliases_visited[k]) {
        return;
    }
    options.aliases_visited[k] = 1;
    // put datatype into aliases list
    if (node.nodeClass === node_opcua_data_model_1.NodeClass.Variable || node.nodeClass === node_opcua_data_model_1.NodeClass.VariableType) {
        const nodeV = node;
        if (nodeV.dataType && nodeV.dataType.namespace === 0 && nodeV.dataType.value !== 0) {
            // name
            const dataTypeName = b(xw, resolveDataTypeName(addressSpace, nodeV.dataType));
            if (dataTypeName) {
                if (!options.aliases[dataTypeName]) {
                    options.aliases[dataTypeName] = n(xw, nodeV.dataType);
                }
            }
        }
        if (nodeV.dataType && nodeV.dataType.namespace !== 0 && nodeV.dataType.value !== 0) {
            // name
            const dataTypeName = b(xw, resolveDataTypeName(addressSpace, nodeV.dataType));
            if (dataTypeName) {
                if (!options.aliases[dataTypeName]) {
                    options.aliases[dataTypeName] = n(xw, nodeV.dataType);
                }
            }
        }
    }
    function collectReferenceNameInAlias(reference) {
        // reference.referenceType
        const key = b(xw, reference._referenceType.browseName);
        if (!options.aliases.key) {
            if (reference.referenceType.namespace === 0) {
                options.aliases[key] = reference.referenceType.toString().replace("ns=0;", "");
            }
            else {
                options.aliases[key] = n(xw, reference.referenceType);
            }
        }
    }
    node.allReferences().forEach(collectReferenceNameInAlias);
}
function writeAliases(xw, aliases) {
    xw.startElement("Aliases");
    if (aliases) {
        const keys = Object.keys(aliases).sort();
        for (const key of keys) {
            xw.startElement("Alias");
            xw.writeAttribute("Alias", key);
            xw.text(aliases[key].toString().replace(/ns=0;/, ""));
            xw.endElement();
        }
    }
    xw.endElement();
}
function constructNamespaceTranslationTable(dependency) {
    const translationTable = {};
    for (let i = 0; i < dependency.length; i++) {
        translationTable[dependency[i].index] = i;
    }
    return translationTable;
}
function dumpReferenceType(xw, referenceType) {
    _markAsVisited(xw, referenceType);
    xw.startElement("UAReferenceType");
    dumpCommonAttributes(xw, referenceType);
    dumpCommonElements(xw, referenceType);
    if (referenceType.inverseName /* LocalizedText*/) {
        xw.startElement("InverseName");
        xw.text(referenceType.inverseName.text || "");
        xw.endElement();
    }
    xw.endElement();
}
function sortByBrowseName(x, y) {
    const x_str = x.browseName.toString();
    const y_str = y.browseName.toString();
    if (x_str > y_str) {
        return -1;
    }
    else if (x_str < y_str) {
        return 1;
    }
    return 0;
}
function dumpXml(node, options) {
    const namespace = node.namespace;
    // make a first visit so that we determine which node to output and in which order
    const nodesToVisit = {};
    const dependency = construct_namespace_dependency_1.constructNamespaceDependency(namespace);
    const translationTable = constructNamespaceTranslationTable(dependency);
    const xw = new XMLWriter(true);
    xw.translationTable = translationTable;
    visitUANode(node, nodesToVisit, false);
    xw.startDocument({ encoding: "utf-8" });
    xw.startElement("UANodeSet");
    xw.writeAttribute("xmlns:xs", "http://www.w3.org/2001/XMLSchema-instance");
    xw.writeAttribute("xmlns:xsd", "http://www.w3.org/2001/XMLSchema");
    xw.writeAttribute("Version", "1.02");
    xw.writeAttribute("LastModified", new Date().toISOString());
    xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2011/03/UANodeSet.xsd");
    buildUpAliases(node, xw, nodesToVisit);
    writeAliases(xw, nodesToVisit.aliases);
    for (const el of nodesToVisit.elements) {
        el.dumpXML(xw);
    }
    xw.endElement();
    xw.endDocument();
    return xw.toString();
}
exports.dumpXml = dumpXml;
namespace_1.UANamespace.prototype.toNodeset2XML = function () {
    const dependency = construct_namespace_dependency_1.constructNamespaceDependency(this);
    const translationTable = constructNamespaceTranslationTable(dependency);
    const xw = new XMLWriter(true);
    xw.translationTable = translationTable;
    xw.startDocument({ encoding: "utf-8", version: "1.0" });
    xw.startElement("UANodeSet");
    xw.writeAttribute("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance");
    xw.writeAttribute("xmlns:uax", "http://opcfoundation.org/UA/2008/02/Types.xsd");
    xw.writeAttribute("xmlns", "http://opcfoundation.org/UA/2011/03/UANodeSet.xsd");
    // xx xw.writeAttribute("Version", "1.02");
    // xx xw.writeAttribute("LastModified", (new Date()).toISOString());
    // ------------- Namespace Uris
    xw.startElement("NamespaceUris");
    // xx const namespaceArray = namespace.addressSpace.getNamespaceArray();
    for (const depend of dependency) {
        if (depend.index === 0) {
            continue; // ignore namespace 0
        }
        xw.startElement("Uri");
        xw.text(depend.namespaceUri);
        xw.endElement();
    }
    xw.endElement();
    // ------------- Namespace Uris
    xw.startElement("Models");
    xw.endElement();
    const s = {};
    for (const node of this.nodeIterator()) {
        buildUpAliases(node, xw, s);
    }
    writeAliases(xw, s.aliases);
    xw.visitedNode = {};
    // -------------- writeReferences
    xw.writeComment("ReferenceTypes");
    const referenceTypes = [...this._referenceTypeIterator()].sort(sortByBrowseName);
    for (const referenceType of referenceTypes) {
        dumpReferenceType(xw, referenceType);
    }
    // -------------- Dictionaries
    const addressSpace = this.addressSpace;
    const opcBinaryTypeSystem = addressSpace.findNode("OPCBinarySchema_TypeSystem");
    if (opcBinaryTypeSystem) {
        // let find all DataType dictionary node corresponding to a given namespace
        // (have DataTypeDictionaryType)
        const nodeToBrowse = new node_opcua_types_1.BrowseDescription({
            browseDirection: node_opcua_data_model_1.BrowseDirection.Forward,
            includeSubtypes: false,
            nodeClassMask: node_opcua_data_model_1.makeNodeClassMask("Variable"),
            nodeId: opcBinaryTypeSystem.nodeId,
            referenceTypeId: node_opcua_nodeid_1.resolveNodeId("HasComponent"),
            resultMask: node_opcua_data_model_1.makeResultMask("ReferenceType | IsForward | BrowseName | NodeClass | TypeDefinition")
        });
        const result = opcBinaryTypeSystem.browseNode(nodeToBrowse).filter((r) => r.nodeId.namespace === this.index);
        node_opcua_assert_1.assert(result.length <= 1);
        if (result.length === 1) {
            xw.writeComment("DataSystem");
            const dataSystemType = addressSpace.findNode(result[0].nodeId);
            dataSystemType.dumpXML(xw);
        }
    }
    // -------------- DataTypes
    const dataTypes = [...this._dataTypeIterator()].sort(sortByBrowseName);
    if (dataTypes.length) {
        xw.writeComment("DataTypes");
        // xx xw.writeComment(" "+ objectTypes.map(x=>x.browseName.name.toString()).join(" "));
        for (const dataType of dataTypes) {
            if (!xw.visitedNode[_hash(dataType)]) {
                dataType.dumpXML(xw);
            }
        }
    }
    // -------------- ObjectTypes
    xw.writeComment("ObjectTypes");
    const objectTypes = [...this._objectTypeIterator()].sort(sortByBrowseName);
    // xx xw.writeComment(" "+ objectTypes.map(x=>x.browseName.name.toString()).join(" "));
    for (const objectType of objectTypes) {
        if (!xw.visitedNode[_hash(objectType)]) {
            objectType.dumpXML(xw);
        }
    }
    // -------------- VariableTypes
    xw.writeComment("VariableTypes");
    const variableTypes = [...this._variableTypeIterator()].sort(sortByBrowseName);
    // xx xw.writeComment("ObjectTypes "+ variableTypes.map(x=>x.browseName.name.toString()).join(" "));
    for (const variableType of variableTypes) {
        if (!xw.visitedNode[_hash(variableType)]) {
            variableType.dumpXML(xw);
        }
    }
    // -------------- Any   thing else
    xw.writeComment("Other Nodes");
    const nodes = [...this.nodeIterator()].sort(sortByBrowseName);
    for (const node of nodes) {
        if (!xw.visitedNode[_hash(node)]) {
            node.dumpXML(xw);
        }
    }
    xw.endElement();
    xw.endDocument();
    return xw.toString();
};
