// node-opcua-pki/dist/misc/subject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Subject = void 0;
const _keys = {
    C: "country",
    CN: "commonName",
    DC: "domainComponent",
    L: "locality",
    O: "organization",
    OU: "organizationalUnit",
    ST: "state",
};
class Subject {
    constructor(options) {
        if (typeof (options) === "string") {
            options = Subject.parse(options);
        }
        this.commonName = options.commonName;
        this.organization = options.organization;
        this.organizationalUnit = options.organizationalUnit;
        this.locality = options.locality;
        this.state = options.state;
        this.country = options.country;
        this.domainComponent = options.domainComponent;
    }
    static parse(str) {
        const elements = str.split("/");
        const options = {};
        elements.forEach((element) => {
            if (element.length === 0) {
                return;
            }
            const s = element.split("=");
            if (s.length !== 2) {
                throw new Error("invalid format for " + element);
            }
            const longName = _keys[s[0]];
            const value = s[1];
            options[longName] = Buffer.from(value, "ascii").toString("utf8");
        });
        return options;
    }
    toString() {
        let tmp = "";
        if (this.country) {
            tmp += "/C=" + this.country;
        }
        if (this.state) {
            tmp += "/ST=" + this.state;
        }
        if (this.locality) {
            tmp += "/L=" + this.locality;
        }
        if (this.organization) {
            tmp += "/O=" + this.organization;
        }
        if (this.organizationalUnit) {
            tmp += "/OU=" + this.organization;
        }
        if (this.commonName) {
            tmp += "/CN=" + this.commonName;
        }
        if (this.domainComponent) {
            tmp += "/DC=" + this.domainComponent;
        }
        return tmp;
    }
}
exports.Subject = Subject;
