// node-opcua-service-filter/dist/tools_event_filter.js
"use strict";
/**
 * @module node-opcua-service-filter
 */
// tslint:disable:object-literal-shorthand
// tslint:disable:only-arrow-functions
// tslint:disable:max-line-length
Object.defineProperty(exports, "__esModule", { value: true });
exports.simpleAttributeOperandToShortString = exports.constructEventFilter = void 0;
const node_opcua_constants_1 = require("node-opcua-constants");
const node_opcua_data_model_1 = require("node-opcua-data-model");
const node_opcua_debug_1 = require("node-opcua-debug");
const node_opcua_nodeid_1 = require("node-opcua-nodeid");
const imports_1 = require("./imports");
const debugLog = node_opcua_debug_1.make_debugLog(__filename);
const doDebug = node_opcua_debug_1.checkDebugFlag(__filename);
/**
 * helper to construct event filters:
 * construct a simple event filter
 *
 *
 * @example
 *
 *     constructEventFilter(["SourceName","Message","ReceiveTime"]);
 *
 *     constructEventFilter(["SourceName",{namespaceIndex:2 , "MyData"}]);
 *     constructEventFilter(["SourceName","2:MyData" ]);
 *
 *     constructEventFilter(["SourceName" ,["EnabledState","EffectiveDisplayName"] ]);
 *     constructEventFilter(["SourceName" ,"EnabledState.EffectiveDisplayName" ]);
 *
 */
function constructEventFilter(arrayOfNames, conditionTypes) {
    if (!Array.isArray(arrayOfNames)) {
        return constructEventFilter([arrayOfNames], conditionTypes);
    }
    if (conditionTypes && !Array.isArray(conditionTypes)) {
        return constructEventFilter(arrayOfNames, [conditionTypes]);
    }
    // istanbul ignore next
    if (!(arrayOfNames instanceof Array)) {
        throw new Error("internal error");
    }
    // replace "string" element in the form A.B.C into [ "A","B","C"]
    const arrayOfNames2 = arrayOfNames.map((path) => {
        if (typeof path !== "string") {
            return path;
        }
        return path.split(".");
    });
    const arrayOfNames3 = arrayOfNames2.map((path) => {
        if (Array.isArray(path)) {
            return path.map(node_opcua_data_model_1.stringToQualifiedName);
        }
        return path;
    });
    // replace "string" elements in arrayOfName with QualifiedName in namespace 0
    const arrayOfNames4 = arrayOfNames3.map((s) => {
        return typeof s === "string" ? node_opcua_data_model_1.stringToQualifiedName(s) : s;
    });
    // construct browse paths array
    const browsePaths = arrayOfNames4.map((s) => {
        return Array.isArray(s) ? s : [s];
    });
    // Part 4 page 127:
    // In some cases the same BrowsePath will apply to multiple EventTypes. If the Client specifies the BaseEventType
    // in the SimpleAttributeOperand then the Server shall evaluate the BrowsePath without considering the Type.
    // [..]
    // The SimpleAttributeOperand structure allows the Client to specify any Attribute, however, the Server is only
    // required to support the Value Attribute for Variable Nodes and the NodeId Attribute for Object Nodes.
    // That said, profiles defined in Part 7 may make support for additional Attributes mandatory.
    let selectClauses = browsePaths.map((browsePath) => {
        return new imports_1.SimpleAttributeOperand({
            attributeId: node_opcua_data_model_1.AttributeIds.Value,
            browsePath,
            indexRange: undefined,
            typeDefinitionId: node_opcua_nodeid_1.makeNodeId(node_opcua_constants_1.ObjectTypeIds.BaseEventType) // i=2041
        });
    });
    if (conditionTypes && conditionTypes instanceof Array) {
        const extraSelectClauses = conditionTypes.map((nodeId) => {
            return new imports_1.SimpleAttributeOperand({
                attributeId: node_opcua_data_model_1.AttributeIds.NodeId,
                browsePath: undefined,
                indexRange: undefined,
                typeDefinitionId: nodeId // conditionType for instance
            });
        });
        selectClauses = selectClauses.concat(extraSelectClauses);
    }
    const filter = new imports_1.EventFilter({
        selectClauses: selectClauses,
        whereClause: {
            // ContentFilter
            elements: [
            // ContentFilterElement
            // {
            //    filterOperator: FilterOperator.IsNull,
            //    filterOperands: [ //
            //        new ElementOperand({
            //            index: 123
            //        }),
            //        new AttributeOperand({
            //            nodeId: "i=10",
            //            alias: "someText",
            //            browsePath: { //RelativePath
            //
            //            },
            //            attributeId: AttributeIds.Value
            //        })
            //    ]
            // }
            ]
        }
    });
    return filter;
}
exports.constructEventFilter = constructEventFilter;
/**
 * @class SimpleAttributeOperand
 * @method toPath
 * @return {String}
 *
 * @example:
 *
 *
 */
function simpleAttributeOperandToPath(self) {
    if (!self.browsePath) {
        return "";
    }
    return self.browsePath
        .map((a) => {
        return a.name;
    })
        .join("/");
}
/**
 * @class SimpleAttributeOperand
 * @method toShortString
 * @return {String}
 *
 * @example:
 *
 *
 */
function simpleAttributeOperandToShortString(self, addressSpace // Address Space
) {
    let str = "";
    if (addressSpace) {
        const n = addressSpace.findNode(self.typeDefinitionId);
        str += n.BrowseName.toString();
    }
    str += "[" + self.typeDefinitionId.toString() + "]" + simpleAttributeOperandToPath(self);
    return str;
}
exports.simpleAttributeOperandToShortString = simpleAttributeOperandToShortString;
