// node-opcua-server/dist/i_server_side_publish_engine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferedSubscription = void 0;
const node_opcua_types_1 = require("node-opcua-types");
const node_opcua_assert_1 = require("node-opcua-assert");
class TransferedSubscription {
    constructor(options) {
        this._pending_notifications = [];
        this.id = options.id;
        this._sequence_number_generator = options.generator;
        this.publishEngine = options.publishEngine;
    }
    get hasPendingNotifications() {
        return this._pending_notifications.length > 0;
    }
    dispose() {
        this._pending_notifications = [];
        this.publishEngine = null;
    }
    _publish_pending_notifications() {
        const n = this._pending_notifications.shift();
        const notificationMessage = n.notification;
        const moreNotifications = false;
        const subscriptionId = this.id;
        node_opcua_assert_1.default(notificationMessage.notificationData.length > 0);
        // ---------------------- to do : duplicated code
        const response = new node_opcua_types_1.PublishResponse({
            moreNotifications,
            notificationMessage: {
                notificationData: notificationMessage.notificationData,
                sequenceNumber: notificationMessage.sequenceNumber
            },
            subscriptionId
        });
        // apply sequence number and store in sent_notifications queue
        node_opcua_assert_1.default(notificationMessage.sequenceNumber === 0xffffffff);
        response.notificationMessage.sequenceNumber = this._get_next_sequence_number();
        //xxx    this._sent_notifications.push(response.notificationMessage);
        // get available sequence number;
        const availableSequenceNumbers = [response.notificationMessage.sequenceNumber];
        node_opcua_assert_1.default(!response.notificationMessage ||
            availableSequenceNumbers[availableSequenceNumbers.length - 1] === response.notificationMessage.sequenceNumber);
        response.availableSequenceNumbers = availableSequenceNumbers;
        this.publishEngine._send_response(this, response);
    }
    _get_next_sequence_number() {
        return this._sequence_number_generator ? this._sequence_number_generator.next() : 0;
    }
}
exports.TransferedSubscription = TransferedSubscription;
