// node-opcua-server/dist/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
// tslint:disable-next-line:no-var-requires
const Dequeue = require("dequeue");
class Queue {
    constructor() {
        this._d = new Dequeue();
        this.size = 0;
    }
    clear() {
        this.size = 0;
    }
    shift() {
        if (this.size === 0) {
            return undefined;
        }
        this.size -= 1;
        return this._d.shift();
    }
    push(value) {
        this.size += 1;
        this._d.push(value);
    }
    values() {
        const self = this;
        const iteratable = {
            [Symbol.iterator]() {
                let cursor = self._d.head;
                const iterator = {
                    next() {
                        cursor = cursor.next;
                        if (cursor === self._d.head) {
                            return { done: true, value: null };
                        }
                        const ret = {
                            done: false,
                            value: cursor.data
                        };
                        return ret;
                    }
                };
                return iterator;
            }
        };
        return iteratable;
    }
}
exports.Queue = Queue;
